kvserver: eagerly initialize Raft groups

erikgrinaker opened the pull request:
This patch eagerly initializes Raft groups during replica construction, instead of deferring it until first use. The original motivation for lazy initialization was to avoid election storms, but since replicas start out quiesced they won't do anything until unquiesced anyway, and Raft pre-vote will prevent disturbing established leaders.

This has a minor impact on startup time: node startup time with 50.000 replicas and warm OS caches increased from 12.6 to 13.6 seconds. That seems fine, and it's likely better to do that work before joining the cluster.

This does not yet enforce the invariant that every replica has a non-nil Raft group: destroyed replicas remove their Raft group (releasing its memory for garbage collection), and many tests construct replicas without Raft groups. This may be addressed later.

Resolves #73715.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/104657)
<!-- Reviewable:end -->

erikgrinaker commented:
Seeing some worrying test failures here. Not sure this is worth the risk/effort right now.

```
=== RUN   TestPause_alter_table_add_foreign_key/pause_stage_1_of_2
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1  slice[12,13) out of bound [11,11]
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !goroutine 27557 [running]:
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !runtime/debug.Stack()
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  GOROOT/src/runtime/debug/stack.go:24 +0x65
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !github.com/cockroachdb/cockroach/pkg/util/log.(*loggerT).outputLogEntry(0xc004adaf00, {{{0xc005515200, 0x24}, {0x4e80c77, 0x1}, {0x4e80c77, 0x1}, {0x4e80c77, 0x1}}, 0x1768ea260bbe72ee, ...})
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  github.com/cockroachdb/cockroach/pkg/util/log/clog.go:261 +0xb8
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !github.com/cockroachdb/cockroach/pkg/util/log.logfDepthInternal({0x616ac60, 0xc005c59bc0}, 0x3, 0x4, 0x0, 0x0?, {0x4e71ac3, 0x21}, {0xc004d00440, 0x4, ...})
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  github.com/cockroachdb/cockroach/pkg/util/log/channels.go:106 +0x645
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !github.com/cockroachdb/cockroach/pkg/util/log.logfDepth(...)
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  github.com/cockroachdb/cockroach/pkg/util/log/channels.go:39
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !github.com/cockroachdb/cockroach/pkg/util/log.FatalfDepth(...)
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  github.com/cockroachdb/cockroach/bazel-out/k8-fastbuild/bin/pkg/util/log/log_channels_generated.go:920
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftLogger).Panicf(0xc00bd6eb50, {0x4e71ac3, 0x21}, {0xc004d00440, 0x4, 0x4})
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/raft.go:118 +0xcd
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !go.etcd.io/raft/v3.(*raftLog).mustCheckOutOfBounds(0xc00bfa9220, 0xc, 0xd)
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  go.etcd.io/raft/v3/external/io_etcd_go_raft_v3/log.go:554 +0x23d
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !go.etcd.io/raft/v3.(*raftLog).slice(0xc00bfa9220, 0xc, 0xd, 0x4000000)
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  go.etcd.io/raft/v3/external/io_etcd_go_raft_v3/log.go:492 +0x4e
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !go.etcd.io/raft/v3.(*raftLog).nextCommittedEnts(0xc00bfa9220, 0x0?)
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  go.etcd.io/raft/v3/external/io_etcd_go_raft_v3/log.go:235 +0x19f
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !go.etcd.io/raft/v3.(*RawNode).readyWithoutAccept(_)
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  go.etcd.io/raft/v3/external/io_etcd_go_raft_v3/rawnode.go:146 +0xd7
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !go.etcd.io/raft/v3.(*RawNode).Ready(_)
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  go.etcd.io/raft/v3/external/io_etcd_go_raft_v3/rawnode.go:134 +0x5e
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked.func2(0xc000b27900?)
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:859 +0x115
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).withRaftGroupLocked.func1(...)
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:2063
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).withRaftGroupLocked(0xc00a227200, 0x616ac60?)
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:2064 +0x4e
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:851 +0x2d7
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:807 +0x19b
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processReady(0xc00a0cca80, 0xc0059f9e60?)
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:646 +0x145
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker(0xc003aa3650, {0x616ac60, 0xc006c6cfc0}, {0x6150df0, 0xc00a0cca80}, 0xc00a0cc000)
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:418 +0x1f2
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2({0x616ac60?, 0xc006c6cfc0?})
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:321 +0x45
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x146
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !  github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x43b
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !
F230615 18:58:29.576938 27557 go.etcd.io/raft/v3/log.go:554  [T1,n1,s1,r50/1:/Table/{48-50}] 1 !For more context, check log files in: /artifacts/tmp/_tmp/2cb16150a1d62f4db8cc89cc951df360/logTestPause_alter_table_add_foreign_key4248197905
```

erikgrinaker committed:
kvserver: eagerly initialize Raft groups

This patch eagerly initializes Raft groups during replica construction,
instead of deferring it until first use. The original motivation for
lazy initialization was to avoid election storms, but since replicas
start out quiesced they won't do anything until unquiesced anyway, and
Raft pre-vote will prevent disturbing established leaders.

This has a minor impact on startup time: node startup time with 50.000
replicas and warm OS caches increased from 12.6 to 13.6 seconds. That
seems fine, and it's likely better to do that work before joining the
cluster.

This does not yet enforce the invariant that every replica has a non-nil
Raft group: destroyed replicas remove their Raft group (releasing its
memory for garbage collection), and many tests construct replicas
without Raft groups. This may be addressed later.

Epic: none
Release note: None

erikgrinaker commented:
I cut the scope of this down from always enforcing a non-nil Raft group to simply initializing during replica creation, since there were a bunch of tests that relied on creating replicas with no Raft groups.

tbg commented on pkg/kv/kvserver/replica_init.go:
```diff
@@ -71,13 +71,29 @@ func newInitializedReplica(store *Store, loaded kvstorage.LoadedReplicaState) (*
+// newReplica creates a new, bare replica without a Raft group. Use either
+// newInitializedReplica() or newUninitializedReplica() instead, which will
+// include an appropriate Raft group.
+func newReplica(store *Store, rangeID roachpb.RangeID, replicaID roachpb.ReplicaID) *Replica {
```
This seems somewhat worse than before, we already need uninited replicas to have the ability to become initialized, and both are valid concepts whereas what's returned from `newReplica` is essentially a "broken" replica. Were there complications in the old structure that made this `newReplica` necessary?

erikgrinaker replied:
This is an internal helper for `newInitializedReplica()` and `newUninitializedReplica()`. If we have `Initialized` simply call through to `Uninitialized` then we'll unnecessarily create a Raft group twice in rapid succession, because `Initialized` has to replace the Raft group from an uninitialized replica.

We could alternatively pass a boolean parameter to `Uninitialized` controlling whether to create a Raft group, but that seemed even worse since `Uninitialized` is a widely used API, which could lead callers to believe that it was legit to create it with a Raft group.

erikgrinaker replied:
I renamed it to `newUninitializedReplicaWithoutRaftGroup` to try to make this a bit clearer, but didn't find a good way to otherwise avoid this.

erikgrinaker commented on pkg/kv/kvserver/replica_init.go:
```diff
@@ -185,10 +201,6 @@ func (r *Replica) setStartKeyLocked(startKey roachpb.RKey) {
-//
-// This method is called in:
-// - loadInitializedReplicaForTesting, to finalize creating an initialized replica;
-// - splitPostApply, to initialize a previously uninitialized replica.
```
This just seemed stale, and prone to become immediately stale again if I updated it. If people are interested in where it's called from, they can use their IDE to find out.
